### ISSUE TITLE
To do Constraint using Sverchok Object ID

### ISSUE THREAD

**Issue by enzyme69** [nortikin/sverchok#1111]
Probably question for @Kosvor2 

![screen shot 2017-01-11 at 17 42 48](https://cloud.githubusercontent.com/assets/3952687/21838335/918197b0-d825-11e6-8509-1435a0771037.png)

If I were to do Constraining of an Object, do I set it up like above?

Maybe it's more like Drivers?

Wonder why when I use Math it does not work.

The SET node is like hybrid get and set value, but sometimes I don't get the value.

**Comment by zeffii** [comment]
the Math node (vector math node) is limiting the socket input to VerticesSocket only. I rewrote VectorMath node yesterday and removed that limitation. -- meaning, When I push the VectorMath node MK2 (soon) this will work properly.

**Comment by zeffii** [comment]
secondly, I have a few issues with the node and some usability enhancements in mind..

- I think the Object ID Set node should probably switch socket types automatically, like the Get/Set prop node
- User could have a choice between using a dropdown (enum) of predecided properties (location, scale, rotation_euler, matrix_world) or the String entry box. (if no string is written then it should show the enum)

**Comment by zeffii** [comment]
![kbar](https://cloud.githubusercontent.com/assets/619340/21845436/f1e02c44-d7f2-11e6-924e-588ed133994d.gif)

**Comment by zeffii** [comment]
![ezbos](https://cloud.githubusercontent.com/assets/619340/21846241/8b84f944-d7f6-11e6-8f44-009119ba24a0.gif)

**Comment by ly29** [comment]
Looks nice

**Comment by zeffii** [comment]
```python
    def process(self):
        O, V = self.inputs
        Ov = self.outputs[0]
        Prop = self.formula or self.selected_mode
        objs = O.sv_get()
        if isinstance(objs[0], list):
            if V.is_linked:
                v = V.sv_get()
                if isinstance(v[0], list):
                    objs, v = safc(objs, v)
                else:
                    objs, v = safc(objs, [v])
                for OBL, VALL in zip(objs, v):
                    OBL, VALL = safc(OBL, VALL)
                    exec("for i, i2 in zip(OBL, VALL):\n    i."+Prop+"= i2")
            elif Ov.is_linked:
                Ov.sv_set(eval("[[i."+Prop+" for i in OBL] for OBL in objs]"))
            else:
                exec("for OL in objs:\n    for i in OL:\n        i."+Prop)
        else:
            if V.is_linked:
                v = V.sv_get()
                if isinstance(v[0], list):
                    v = v[0]
                objs, v = safc(objs, v)
                exec("for i, i2 in zip(objs, v):\n    i."+Prop+"= i2")
            elif Ov.is_linked:
                Ov.sv_set(eval("[i."+Prop+" for i in objs]"))
            else:
                exec("for i in objs:\n    i."+Prop)
```

I don't want to do too much to this node.. but the style here bothers me.

**Comment by ly29** [comment]
Well it bothers me as well, both the variable names and the execs/evals.

**Comment by zeffii** [comment]
rewriting now :)

**Comment by zeffii** [comment]
first pass
```
    # if prop has a dot then   operator.attrgetter('b.c')(a)

    def process(self):
        (O, V), Ov = self.inputs, self.outputs
        Prop = self.formula or self.selected_mode

        objs = O.sv_get()
        v = V.sv_get()

        if isinstance(objs[0], list):
            if V.is_linked:
                v = v if isinstance(v[0], list) else [v]
                objs, v = safc(objs, v)

                for OBL, VALL in zip(objs, v):
                    OBL, VALL = safc(OBL, VALL)
                    for i, i2 in zip(OBL, VALL):
                        setattr(i, Prop, i2)

            elif Ov.is_linked:
                Ov.sv_set([[getattr(i, Prop) for i in OBL] for OBL in objs])
            else:
                pass  # as far as i can tell without executing, this wasn't doing anything.

        else:
            if V.is_linked:
                v = v[0] if isinstance(v[0], list) else v
                objs, v = safc(objs, v)

                for i, i2 in zip(objs, v):
                    setattr(i, Prop, i2)

            elif Ov.is_linked:
                Ov.sv_set([getattr(i, Prop) for i in objs])
            else:
                pass # as far as i can tell without executing, this wasn't doing anything.
```

**Comment by zeffii** [comment]
first pass untested : https://gist.github.com/zeffii/6907d11bb4f3058642ce5b5507f119eb

**Comment by Kosvor2** [comment]
@zeffii does `getattr(i, Prop)` acept Props with '.' like `location.z` or `data.name` ? if not, dont use it.
Node also must work when no Value socket is used -see https://github.com/nortikin/sverchok/issues/1067

![bc861100-cf5d-11e6-90bf-635f073a24c7](https://cloud.githubusercontent.com/assets/22656834/21847983/b5e50c56-d80f-11e6-8eb0-2d0e78796b69.png)

**Comment by zeffii** [comment]
yep, prop with dot should work now.

**Comment by zeffii** [comment]
(well, it will.. just not yet )

**Comment by ly29** [comment]
I am not certain that I agree about the without values part but sure that could be done.

**Comment by zeffii** [comment]
see @Kosvor2    https://docs.python.org/3.5/library/operator.html#operator.attrgetter

**Comment by zeffii** [comment]
@Kosvor2  if my modifications don't reach feature parity, don't worry I wont push :)

**Comment by zeffii** [comment]
ok i see,
```python
            else:
                exec("for OL in objs:\n    for i in OL:\n        i."+Prop)
```
that's what executes the  `i.prop = something` case.  then let's do a `=` lookup for that case.

**Comment by ly29** [comment]
It is just that seeing ```exec``` and ```eval``` makes me itchy.

**Comment by zeffii** [comment]
all eval / exec can be avoided in this node.

**Comment by zeffii** [comment]
indeed, it can probably been avoided in all nodes... but this one is easy to modify

**Comment by ly29** [comment]
Why is bad? Because it is hard to follow the code and see what it does.

Well in some makes sense, like the script nodes and formula but otherwise it is better avoided.

**Comment by Kosvor2** [comment]
@enzyme69 you forgot to use `[:]` to convert mathutils data to tuples.

@zeffii , @ly29 manual has some gud images for compare new implenentation to: https://github.com/nortikin/sverchok/blob/master/docs/nodes/beta_nodes/set_dataobject.rst

**Comment by ly29** [comment]
@Kosvor2 
Some interesting examples indeed. Some feels more like side effects than designed.
The variable names ```i``` and ```i2``` for object and value are also something I missed from reading the code.

How is # used?

**Comment by zeffii** [comment]
retain all existing behaviour is proving difficult. i'll have to suspend my attempt :)

**Comment by nortikin** [comment]
how to explain user why we use in one case n[0], n[1], n[k]... in other U,V in other A,B in other i, i2 etc?
Specially not obvious if we use not observed in sockets veriables.

**Comment by Kosvor2** [comment]
@ly29 `#` is doing its work inside of the node, see https://github.com/nortikin/sverchok/issues/1067

**Comment by ly29** [comment]
So basically as a comment, ok.
